### PR TITLE
GI-105 adding JavaScript to the vote button on show idea page

### PR DESCRIPTION
**Pull request by Pareis2**
- Added JS to the vote button element, which would increment the vote number by 1 and then disable the button.
- Created an error message if vote has failed
- Display remove button if user has already voted
- Added authenticity token for cross site scripting

### CODE REVIEW

File: app/assets/stylesheets/forms.scss [hunk 1]
```diff
@@ -17,3 +20,9 @@ div#votes div {
 div#votes{
   text-align: right;
 }
+div#votes p#voteerror {
+  display: block;
```
Comment (kmahern) [inline app/assets/stylesheets/forms.scss]: Won't these be `display: block` already?

File: app/assets/stylesheets/forms.scss [hunk 2]
```diff
@@ -17,3 +20,9 @@ div#votes div {
...
+
+  color: #B10E1E;
```
Comment (kmahern) [inline app/assets/stylesheets/forms.scss]: This looks very similar to the `.govuk-error-summary.alert` rule.  Could we re-use this one?

File: app/controllers/votes_controller.rb [hunk 3]
```diff
@@ -12,7 +12,8 @@ def create
     @vote = @idea.votes.build
     @vote.user = current_user
     @vote.save
-    redirect_to @idea, notice: 'Your vote has been added'
+    # redirect_to @idea, notice: 'Your vote has been added'
```
Comment (kmahern) [inline app/controllers/votes_controller.rb]: Is this code intentionally commented?

File: app/views/ideas/show.html.erb [hunk 4]
```diff
@@ -61,12 +61,12 @@
 <div id="votes" class="column-one-third">
   <% if @idea.approved_by_admin? %>
     <p class="govuk-body-l">Votes</p>
-    <p class="govuk-body-l"><%= @idea.votes.count %></p>
+    <p id="vcount" class="govuk-body-l"><%= @idea.votes.count %></p>
```
Comment (kmahern) [inline app/views/ideas/show.html.erb]: Minor, but I was a little confused by `vcount`.  I'd prefer a non-shortened vote.

File: app/views/ideas/show.html.erb [hunk 5]
```diff
@@ -61,12 +61,12 @@
...
     <div id="votebutton">
       <% if @idea.user_voted?(current_user) %>
-        <%= button_to 'Remove Vote', idea_vote_path(@idea, @vote.id), method: :delete, class: "govuk-button" %>
+        <%= button_to 'Remove Vote', idea_vote_path(@idea, @vote.id), method: :delete, class: "govuk-button", id: "removevote" %>
```
Comment (kmahern) [inline app/views/ideas/show.html.erb]: Can we have a consistent naming system for HTML ids?
Taking a look at the repo I see both "an-id" and "an_id" being used.

We've taken a look at the Apply code, and they're using "-"s to separate words.  We'd vote for that. 👍

File: app/views/ideas/show.html.erb [hunk 6]
```diff
@@ -81,3 +81,50 @@
 <% end %>
 <div class="govuk-!-padding-bottom-5"></div>
 <%= link_to 'Back', ideas_path %>
+
+<script>
```
Comment (kmahern) [inline app/views/ideas/show.html.erb]: It would be nice to extract this out into a JS file, however I'd view this as minor issue for now.

File: app/views/ideas/show.html.erb [hunk 7]
```diff
@@ -81,3 +81,50 @@
...
+  function vote(event) {
```
Comment (kmahern) [inline app/views/ideas/show.html.erb]: I find the indentation inside this method hard to follow.  Do you have an auto-formatter configured which can fix this for you?  If not give me a shout and I can show you.

Comment (kmahern) [reply app/views/ideas/show.html.erb]: The vote function is quite long.  It would be good to investigate refactoring some of the functionality into smaller methods.

File: app/views/ideas/show.html.erb [hunk 8]
```diff
@@ -81,3 +81,50 @@
...
+    event.preventDefault();
+
+    var token = document.querySelector('meta[name="csrf-token"]').content
+
+    let xhttp = new XMLHttpRequest();
```
Comment (kmahern) [inline app/views/ideas/show.html.erb]: Minor point, you've mixed `var` and `let` in this method.

File: app/controllers/votes_controller.rb (hunk 3 above)
Comment (heidar) [reply app/controllers/votes_controller.rb]: This needs a ruby dev to look at these controller actions to make them work with and without the JS.
